src/main/java/geocode/PostalCode.java

```diff
@@ -0,0 +1,130 @@
+/*
+The MIT License (MIT)
+[OSI Approved License]
+The MIT License (MIT)
+Copyright (c) 2014 Daniel Glasson
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+package geocode;
+
+import static java.lang.Math.cos;
+import static java.lang.Math.sin;
+import static java.lang.Math.toRadians;
+
+import geocode.kdtree.KDNodeComparator;
+
+import java.util.Comparator;
+
+/**
+ * This class works with a postal code file from http://download.geonames.org/export/zip/
+ */
+
+public class PostalCode extends KDNodeComparator<PostalCode> {
+    public String countryCode;
+    public String postalCode;
+    public String placeName;
+    public String adminName1;
+    public String adminCode1;
+    public String adminName2;
+    public String adminCode2;
+    public String adminName3;
+    public String adminCode3;
+    public double latitude;
+    public double longitude;
+    public int accuracy;
+    public double[] point = new double[3]; // The 3D coordinates of the point
+
+    PostalCode(String row) {
+        String[] data = row.split("\t");
+        countryCode = data[0];
+        postalCode = data[1];
+        placeName = data[2];
+        adminName1 = data[3];
+        adminCode1 = data[4];
+        adminName2 = data[5];
+        adminCode2 = data[6];
+        adminName3 = data[7];
+        adminCode3 = data[8];
+        latitude = Double.parseDouble(data[9]);
+        longitude = Double.parseDouble(data[10]);
+        setPoint();
+
+        // Accuracy not always provided
+        if (data.length == 12) {
+            accuracy = Integer.parseInt(data[11]);
+        }
+    }
+
+    PostalCode(Double latitude, Double longitude) {
+        this.placeName = "Search";
+        this.latitude = latitude;
+        this.longitude = longitude;
+        setPoint();
+    }
+
+    private void setPoint() {
```

Comment:
This is how the 3D point is created

Reply:
I presume this is some kind of well-established vector algorithm... but I couldn't find the source with a cursory search. What's the objective of this point conversion?

Reply:
Ah is it just trying to normalize our 2d points into 3d vectors?

Reply:
This part of my code is identical to the GeoName.java file, which is already in the library. I couldn't find this exact math anywhere either so I was hoping that @caleboverman can look it over and we can tweak it if we want. I found some distance calculations that were somewhat similar (https://www.movable-type.co.uk/scripts/latlong.html).

Reply:
My java is a little rusty, but it looks like they're using the `squaredDistance` to find the nearest postal code centroid? The haversine method is the preferred geo distance calculation (linked by @dlfajen) and is what we use in the `distance_to_venue` derived attribute.


There's also the distinct possibility that the nearest centroid is not the actual postal code that your lat/lon indicate. A better solution would be using a true reverse geocode API that relies on shapefiles to determine a postal code. That might be a future iteration...

Reply:
Yes, that's correct. We've already discussed that the FTS lat/lon returns a postal code of 64121 instead of 64108, but we felt like that was probably close enough.

Reply:
@caleboverman  - could you point me to the distance_to_venue code? Maybe I can figure out how to implement that here instead.

Reply:
https://github.com/sporting-innovations/datasci-enriched-attr/blob/61204ecd5d05d26379ae3ac15815d51ddba3db33/distance_to_venue/distance_to_venue.py#L12

Also:
https://rosettacode.org/wiki/Haversine_formula#Java

Reply:
Yeah this doesn't look like the same math, right? I guess I didn't put together how these "point" representations are actually used in every location to see if it maps to a formula we understand... but this is the main usage I see:

point 1 is just cos(lat) * cos(long)
point 2 is cos(lat) * sin (long)
point 3 is sin(lat)

Then those raw values are stored, and when we ask for the `squareDistance` between two "3d points" ([x1,y1,z1], [x2,y2,z2]) it does this:

(x1-x2)^2 + (y1-y2)^2 + (z1-z2)^2

So that expands to roughly (for coordinates [l1,lo1],[l2,lo2]):
((cost(l1) * cos(lo1)) - (cos(l2) * cos(lo2)))^2 + ((cos(l1) * sin (lo1)) - (cos(l2) * sin (lo2)))^2 + (sin(l1) - sin(l2))^2

I'm not sure if that expansion is simplifiable to any of the given formulas above...
Mostly, instead these look like normalization functions for vectors iirc ... I'm still just not really understanding where the mapping from the lat/long to 3d "point" representation comes from. The mapping of a 2d point to a 3d represent

Reply:
Oh here we go, lat/long mapped to geodesy vectors : https://www.movable-type.co.uk/scripts/latlong-vectors.html

then it just does normal vector distance calculations

Reply:
Is this distance calculation implemented properly though? My trigonometric function conversions are a big rusty... https://www.movable-type.co.uk/scripts/latlong-vectors.html#distance

> d = R · atan2(|a×b|, a·b)
> where R is the radius of the earth (normally taken as 6371 km).

Reply:
Hmm, yeah... I'm not finding where I found that distance equation. I'll change it to match what you've posted because that link is exactly what this other code is doing.

Reply:
So I changed the distance in the max distance check to use the haversine distance instead. I'm wondering if I should do anything with the other distance checks that are in the KD tree lookups. I think he coded it like this as more of a quick approximation.

Reply:
Yeah my impression was it used a quick check to narrow the results. So it’s probably ok to only use haversine once the search space has been narrowed.

Reply:
Is there a variation of the haversine algorithm for vectors? The (haversine) links above always just use coordinates directly so if you wanted to use that you could skip the vectorization right? 

Otherwise, given the converted vectors you can do the more simple calculations based on the distance between two vectors on a sphere (given the radius) which is the function I quoted above with arctan and the cross product of the vectors.

Reply:
Switched to use the distance calculation using arctan(cross product, dot product). Also added a junit test for it.